[FLINK-11516][table] Port and move some Descriptor classes to flink-table-common

dianfu opened the pull request:
## What is the purpose of the change

*This pull request ported the following descriptor related classes to flink-table-common: Metadata, MetadataValidator, TableStats, ColumnStats, Statistics, StatisticsValidator,TableDescriptor, StreamableDescriptor, StreamTableDescriptorValidator*

## Brief change log

  - *Ported the classes and tests to flink-table-common*
  - *Make StreamableDescriptor not to extends TableDescriptor any more*

## Verifying this change

This change is a code cleanup without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)

dianfu commented:
@twalthr @dawidwys I have split the changes related to Descriptor to this PR. Looking forward to your review. Thanks in advance.

rmetzger commented:
@flinkbot please re-render the tracking comment so that @twalthr sees that this feature request got implemented :)

twalthr commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/descriptors/Metadata.java:
```diff
@@ -0,0 +1,88 @@
+	@Override
+	public final Map<String, String> toProperties() {
+		DescriptorProperties properties = new DescriptorProperties();
+		if (comment != null) {
```
We should avoid null in Java. Use an approach similar to `org.apache.flink.table.descriptors.Csv` or `org.apache.flink.table.descriptors.Elasticsearch` for all descriptors.

twalthr commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/descriptors/StatisticsValidator.java:
```diff
@@ -0,0 +1,127 @@
+		Map<String, ColumnStats> stats = new HashMap<>();
+		for (int i = 0; i < columnCount; i++) {
+			final String propertyKey = key + "." + i + "." + NAME;
+			String name = properties.getOptionalString(propertyKey).orElseThrow(exceptionSupplier(propertyKey));
```
If you use `getString` instead of `getOptionalString`, you don't need the custom exception supplier.

twalthr commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/descriptors/StreamTableDescriptorValidator.java:
```diff
@@ -0,0 +1,65 @@
+ * Validator for {@code StreamTableDescriptor}.
+ */
+@Internal
+public class StreamTableDescriptorValidator implements DescriptorValidator {
```
This should be called `StreamableDescriptorValidator` and the comment updated.

dianfu replied:
This is StreamTableDescriptorValidator actually. :)

twalthr commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/plan/stats/ColumnStats.java:
```diff
@@ -0,0 +1,122 @@
+/**
+ * column statistics.
+ */
+public final class ColumnStats {
```
Make sure to add proper annotations to all classes. This is definitely a `@PublicEvolving`.

twalthr commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/plan/stats/TableStats.java:
```diff
@@ -16,15 +16,40 @@
-case class TableStats(rowCount: Long, colStats: Map[String, ColumnStats] = new HashMap())
+ * Table statistics.
+ */
+public final class TableStats {
```
Add annotation.

twalthr commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/plan/stats/ColumnStats.java:
```diff
@@ -0,0 +1,122 @@
+import java.util.stream.Collectors;
+
+/**
+ * column statistics.
```
nit: We should write the first character of a JavaDoc comment in upper case.

twalthr commented on flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/plan/stats/FlinkStatistic.scala:
```diff
@@ -49,7 +49,7 @@ class FlinkStatistic(tableStats: Option[TableStats]) extends Statistic {
     */
   def getColumnStats(columnName: String): ColumnStats = tableStats match {
-    case Some(tStats) => tStats.colStats.get(columnName)
+    case Some(tStats) => tStats.getColStats.get(columnName)
```
If we change the method names anyway, can we name them properly without abbreviations? `getColumnStats()` or `getMaxValue()` instead of just `getMax()`?

twalthr commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/descriptors/StreamableDescriptor.java:
```diff
@@ -0,0 +1,73 @@
+ * A trait for descriptors that allow to convert between a dynamic table and an external connector.
+ */
+@PublicEvolving
+public interface StreamableDescriptor<D extends StreamableDescriptor<D>> {
```
should extend descriptor

dianfu commented:
@twalthr Thanks a lot for the review. Have updated accordingly with just one exception: StreamTableDescriptorValidator name is not a mistake. :)

godfreyhe commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/descriptors/StatisticsValidator.java:
```diff
@@ -0,0 +1,118 @@
+				properties.getOptionalDouble(key + "." + i + "." + AVG_LENGTH).orElse(null),
+				properties.getOptionalInt(key + "." + i + "." + MAX_LENGTH).orElse(null),
+				properties.getOptionalDouble(key + "." + i + "." + MAX_VALUE).orElse(null),
+				properties.getOptionalDouble(key + "." + i + "." + MIN_VALUE).orElse(null)
```
we should add `max-value-type` property, and get correct value by this property

dianfu commented:
Hi @godfreyhe Thanks a lot for the review. Have addressed your comments except the BigDecimal/BigInteger as it seems that it introduces a new feature to support BigDecimal/BigInteger as the min/max value. What about address that in a separate PR?

godfreyhe commented:
Looks good to me. 
I want to extend `max` and `min` to `Comparable` type (e.g. `Date`, `Time`, `
BigDecimal`) in a new PR.

dianfu commented:
@twalthr @godfreyhe Thanks a lot for the review and commit.
